crates/bevy_ecs/src/schedule/node.rs

```diff
@@ -251,6 +255,186 @@ new_key_type! {
+            NodeId::Set(_) => None,
+        }
+    }
+
+    /// Returns the system set key if the node is a system set, otherwise `None`.
+    pub const fn as_set(&self) -> Option<SystemSetKey> {
+        match self {
+            NodeId::System(_) => None,
+            NodeId::Set(set) => Some(*set),
+        }
+    }
+}
+
+impl GraphNodeId for NodeId {
+    type Adjacent = CompactNodeIdAndDirection;
+    type Edge = CompactNodeIdPair;
+}
+
+impl From<SystemKey> for NodeId {
+    fn from(system: SystemKey) -> Self {
+        NodeId::System(system)
+    }
+}
+
+impl From<SystemSetKey> for NodeId {
+    fn from(set: SystemSetKey) -> Self {
+        NodeId::Set(set)
+    }
+}
+
+/// Compact storage of a [`NodeId`] and a [`Direction`].
+#[derive(Clone, Copy)]
+pub struct CompactNodeIdAndDirection {
+    key: KeyData,
+    is_system: bool,
+    direction: Direction,
+}
+
+impl Debug for CompactNodeIdAndDirection {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let tuple: (_, _) = (*self).into();
+        tuple.fmt(f)
+    }
+}
+
+impl From<(NodeId, Direction)> for CompactNodeIdAndDirection {
+    fn from((id, direction): (NodeId, Direction)) -> Self {
+        let key = match id {
+            NodeId::System(key) => key.data(),
+            NodeId::Set(key) => key.data(),
+        };
+        let is_system = id.is_system();
+
+        Self {
+            key,
+            is_system,
+            direction,
+        }
+    }
+}
+
+impl From<CompactNodeIdAndDirection> for (NodeId, Direction) {
+    fn from(value: CompactNodeIdAndDirection) -> Self {
+        let node = match value.is_system {
+            true => NodeId::System(value.key.into()),
+            false => NodeId::Set(value.key.into()),
+        };
+
+        (node, value.direction)
+    }
+}
+
+/// Compact storage of a [`NodeId`] pair.
+#[derive(Clone, Copy, Hash, PartialEq, Eq)]
+pub struct CompactNodeIdPair {
+    key_a: KeyData,
+    key_b: KeyData,
+    is_system_a: bool,
+    is_system_b: bool,
+}
+
+impl Debug for CompactNodeIdPair {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let tuple: (_, _) = (*self).into();
+        tuple.fmt(f)
+    }
+}
+
+impl From<(NodeId, NodeId)> for CompactNodeIdPair {
+    fn from((a, b): (NodeId, NodeId)) -> Self {
+        let key_a = match a {
+            NodeId::System(index) => index.data(),
+            NodeId::Set(index) => index.data(),
+        };
+        let is_system_a = a.is_system();
+
+        let key_b = match b {
+            NodeId::System(index) => index.data(),
+            NodeId::Set(index) => index.data(),
+        };
+        let is_system_b = b.is_system();
+
+        Self {
+            key_a,
+            key_b,
+            is_system_a,
+            is_system_b,
+        }
+    }
+}
+
+impl From<CompactNodeIdPair> for (NodeId, NodeId) {
+    fn from(value: CompactNodeIdPair) -> Self {
+        let a = match value.is_system_a {
```

Comment:
Oh, I didn't notice the `match bool` before.  It was like that before, and it does seem reasonable here for symmetry with the reverse impl, but ... but ... `if`!